tsweb/tsweb.go

```diff
@@ -455,26 +455,49 @@ func WritePrometheusExpvar(w io.Writer, kv expvar.KeyValue) {
 	writePromExpVar(w, "", kv)
 }
 
-func writePromExpVar(w io.Writer, prefix string, kv expvar.KeyValue) {
-	key := kv.Key
+type prometheusMetricDetails struct {
+	Name  string
+	Type  string
+	Label string
+}
+
+var prometheusMetricCache sync.Map // string => *prometheusMetricDetails
+
+func prometheusMetric(prefix string, key string) (string, string, string) {
+	cachekey := prefix + key
+	if v, ok := prometheusMetricCache.Load(cachekey); ok {
+		d := v.(*prometheusMetricDetails)
+		return d.Name, d.Type, d.Label
+	}
 	var typ string
 	var label string
 	switch {
-	case strings.HasPrefix(kv.Key, gaugePrefix):
+	case strings.HasPrefix(key, gaugePrefix):
 		typ = "gauge"
-		key = strings.TrimPrefix(kv.Key, gaugePrefix)
+		key = strings.TrimPrefix(key, gaugePrefix)
 
-	case strings.HasPrefix(kv.Key, counterPrefix):
+	case strings.HasPrefix(key, counterPrefix):
 		typ = "counter"
-		key = strings.TrimPrefix(kv.Key, counterPrefix)
+		key = strings.TrimPrefix(key, counterPrefix)
```

Comment:
I am not quite sure how our prefixing works. It seems to me that we hang if there is not a "gauge" or "counter" prefix, is that correct? does this mean we require all prom metrics to have this prefix?

Reply:
I think the prefixes are optional and there is some logic that attempts to determine metric type based on expvar type, but it's definitely safer to have a prefix.

Longer term we should probably just switch to an open source Prometheus client instead of exposing metrics ourselves via expvar.